group/edwards25519/scalar.go

```
@@ -4,11 +4,196 @@
+	return s
+}
+
+// Set to the modular sum of scalars a and b
+func (s *scalar) Add(a, b kyber.Scalar) kyber.Scalar {
+	scAdd(&s.v, &a.(*scalar).v, &b.(*scalar).v)
+	return s
+}
+
+// Set to the modular difference a - b
+func (s *scalar) Sub(a, b kyber.Scalar) kyber.Scalar {
+	scSub(&s.v, &a.(*scalar).v, &b.(*scalar).v)
+	return s
+}
+
+// Set to the modular negation of scalar a
+func (s *scalar) Neg(a kyber.Scalar) kyber.Scalar {
+	var z scalar
+	z.Zero()
+	scSub(&s.v, &z.v, &a.(*scalar).v)
+	return s
+}
+
+// Set to the modular product of scalars a and b
+func (s *scalar) Mul(a, b kyber.Scalar) kyber.Scalar {
+	scMul(&s.v, &a.(*scalar).v, &b.(*scalar).v)
+	return s
+}
+
+// Set to the modular division of scalar a by scalar b
+func (s *scalar) Div(a, b kyber.Scalar) kyber.Scalar {
+	var i scalar
+	i.Inv(b)
+	scMul(&s.v, &a.(*scalar).v, &i.v)
+	return s
+}
+
+// Set to the modular inverse of scalar a
+func (s *scalar) Inv(a kyber.Scalar) kyber.Scalar {
+	var res scalar
+	res.One()
+	ac := a.(*scalar)
+	// Modular inversion in a multiplicative group is a^(phi(m)-1) = a^-1 mod m
+	// Since m is prime, phi(m) = m - 1 => a^(m-2) = a^-1 mod m.
+	// The inverse is computed  using the exponentation-and-square algorithm.
+	// Implementation is constant time regarding the value a, it only depends on
+	// the modulo.
+	for i := 255; i >= 0; i-- {
+		bit := l_minus_2_big.Bit(i)
+		// square step
+		scMul(&res.v, &res.v, &res.v)
+		if bit == 1 {
+			// multiply step
+			scMul(&res.v, &res.v, &ac.v)
+		}
+	}
+	s.v = res.v
+	return s
+}
+
+// Set to a fresh random or pseudo-random scalar
+func (s *scalar) Pick(rand cipher.Stream) kyber.Scalar {
+	i := mod.NewInt(random.Int(primeOrder, rand), primeOrder)
+	return s.setInt(i)
+}
+
+func (s *scalar) SetBytes(b []byte) kyber.Scalar {
+	// XXX handle simple and scReduce cases appropriately
+	return s.setInt(mod.NewIntBytes(b, primeOrder, mod.LittleEndian))
+}
+
+func (s *scalar) SetVarTime(varTime bool) bool {
+	return false
```

I think it should just always silently return false, as per the current code.  This way:

1. a client who wants to test whether a given Point or Scalar implementation supports constant time at all can see if s.SetVarTime(false).SetVarTime(false) returns false (it's constant-time) or true (it only has a variable-time implementation).

2. a client who wants to test whether a constant-time Point or Scalar implementation supports any variable-time optimizations can test the result of s.SetVarTime(true).SetVarTime(true): if false, it's always constant-time; if true, it is either always or configurably variable-time.